src/libsync/syncengine.cpp

```diff
@@ -1301,33 +1306,14 @@ void SyncEngine::restoreOldFiles()
     }
 }
 
-bool SyncEngine::estimateState(QString fn, csync_ftw_type_e t, SyncFileStatus* s)
+SyncFileItem* SyncEngine::findSyncItem(const QString &fileName) const
 {
-    Q_UNUSED(t);
-    QString pat(fn);
-    if( t == CSYNC_FTW_TYPE_DIR && ! fn.endsWith(QLatin1Char('/'))) {
-        pat.append(QLatin1Char('/'));
-    }
-
     Q_FOREACH(const SyncFileItemPtr &item, _syncedItems) {
```

Comment:
Yes this is slow, this is the main reason that caused #3122, which we worked around by reducing the calls.
It's sorted by destination, so binary search will work most of the time, but not for a rename made by the remote, since we will need to match SyncFileItem::_file in this case instead of SyncFileItem::_renameTarget (which will be the value returned by destination().

This patch isn't changing this though, it's as slow as before.